만국박람회 [STEP 3] 현이, Hugh

Hugh-github opened the pull request:
@Wody95 
안녕하세요 우디!
Step03 리뷰도 잘 부탁드립니다. ☺️

### 구현한 기능
+ Parsing에 실패한 데이터를 제외하고 반환하는 로직 구현
    + 파싱에 실패한 데이터를 nil로 처리해 고차함수 filter를 사용해 Entry의 속성이 nil값이 아닌 데이터만 따로 추출하였습니다.

+ 특정 뷰의 화면 고정을 설정
    + `supportedInterfaceOrientations` 을 사용해 MainView를 세로 모드로 고정했습니다.
MainView를 제외한 나머지 화면은 NavigationController의 **UIInterfaceOrientationMask**와 같은 설정이기 때문에 따로 설정해 주지 않았습니다.

+ Dynamic Type 적용
    + 프로젝트에 사용된 Label에 Dynamic Type을 적용했습니다.

### 궁금한 점
+ 특정 뷰의 화면 고정을 설정해 주었지만 MainView가 가로인 상태에서 ItemListView로 전환할 경우 여전히 화면의 방향이 바뀌지 않는 걸 확인했습니다. 혹시 해당 문제를 해결하는 방법이 있는지 궁금합니다.
![](https://i.imgur.com/cvnuOtB.gif)

+ UIImage에는 Dynamic Type을 적용하는지 궁금합니다.

+ UIButton에 Dynamic Type을 적용해 title 크기를 키우면 텍스트가 줄임표(...) 처리가 됩니다. 줄임표 처리가 되는 게 맞는지 궁금합니다.
![](https://i.imgur.com/UeUvNP9.gif)

Hugh-github committed:
feat: MainViewController만 세로모드가 고정되도록 코드 추가

seohyeon2 committed:
refactor: 버튼 타이틀 잘림현상 해결

seohyeon2 committed:
refactor: MARK 주석 추가

Hugh-github committed:
refactor: Label에 Dynamic Type 적용

Hugh-github committed:
refactor : 화면 고정 로직 변경

seohyeon2 committed:
refactor: 정상적인 데이터만 반환하게 Entry 파일에 init 추가

seohyeon2 committed:
refactor: entryParse 메서드에 테스트코드 반영

Wody95 commented on Expo1900/Expo1900/Controller/NavigationViewController.swift:
```diff
@@ -0,0 +1,32 @@
+
+class NavigationViewController: UINavigationController {
+    override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
+        return self.topViewController!.supportedInterfaceOrientations
```
강제 언레핑!

현 프로젝트에선 topView가 보장될 수 있지만, 강제 언레핑은 되도록 사용하지 않는게 좋습니다.

seohyeon2 replied:
nil coalescing을 이용하여 옵셔널 바인딩 했습니다..!

Wody95 commented on Expo1900/Expo1900Tests/Expo1900Tests.swift:
```diff
@@ -50,11 +50,30 @@ class Expo1900Tests: XCTestCase {
         XCTAssertEqual(["wody", "coda", "현이"], result)
     }
+    
+    func test_잘못된_json데이터가_있을때_정상적인_데이터만_반환해주는지2() {
```
Unit Test 진행하신 점 매우 좋습니다 👏👏👏👏👏

Wody95 reviewed (commented):
먼저 확실한 디자인 가이드 없이 UI 작업 진행하시느라 수고 많으셨습니다.

텍스트에 Dynamic Type을 적용하시면서 [Accessibility](https://developer.apple.com/videos/play/wwdc2019/257/)(접근성) 에 대해 공부도 해보셨을텐데요.
접근성은 기본적으로 모든 사람이 앱을 사용 가능하도록 만드는 기능입니다. (시각 및 청각 장애인들을 돕는 기능으로 주로 사용되기도 합니다.)

[공식문서](https://developer.apple.com/documentation/objectivec/nsobject/uiaccessibility)에서 볼 수 있는 기능 중 현업에서 주로 많이 사용하는 기능들이 있으니 참고해보시면 좋을 것 같습니다.

- ****[accessibilityAttributedHint](https://developer.apple.com/documentation/objectivec/nsobject/2865913-accessibilityattributedhint)****
- ****[accessibilityAttributedLabel](https://developer.apple.com/documentation/objectivec/nsobject/2865944-accessibilityattributedlabel)****
- ****[UIAccessibilityTraits](https://developer.apple.com/documentation/uikit/uiaccessibilitytraits)****

### **궁금한 점 답변**

- 특정 뷰의 화면 고정을 설정해 주었지만 MainView가 가로인 상태에서 ItemListView로 전환할 경우 여전히 화면의 방향이 바뀌지 않는 걸 확인했습니다. 혹시 해당 문제를 해결하는 방법이 있는지 궁금합니다.
    - 현재 디바이스의 방향을 직접 바꿔줘야 회전 이벤트가 발생하여 화면이 방향이 바뀌는 걸로 확인됩니다. 
    화면의 방향을 현재 디바이스 방향에 맞춰 회전을 명령하는 메서드가 있었지만 현재 사용되지 않음 처리 되었으며, 방향을 바꾸지 않으면 ViewController에서 지원하는 화면 방향 제한이 풀리더라도 화면 회전 체크를 안하는걸로 확인됩니다.
    따라서 이 부분은 저도 조금 더 고민하고 공부해봐야 할 것 같습니다.
- UIImage에는 Dynamic Type을 적용하는지 궁금합니다.
    - 이미지에는 다이나믹 타입을 적용하지 않습니다.
- UIButton에 Dynamic Type을 적용해 title 크기를 키우면 텍스트가 줄임표(...) 처리가 됩니다. 줄임표 처리가 되는 게 맞는지 궁금합니다.
    - UIButton에도 폰트 크기에 따라 number Of line을 변경할 수 있으니 확인해보시면 좋을 것 같습니다. → [공식문서](https://developer.apple.com/documentation/uikit/uibutton/1623992-titlelabel)
    
    
    
# 추가사항
현재 ExpoAssets 폴더 내부에 더 이상 사용하지 않는 imageSet이 있습니다. 현재 앱 내부의 이미지 소스는 Assets를 통해 관리하고 있으니 필요 없는 데이터는 지워주세요

Wody95 commented on Expo1900/Expo1900/Controller/MainViewController.swift:
```diff
@@ -119,6 +135,8 @@ final class MainViewController: UIViewController {
     
+    // MARK: - Method
+    
     @objc func didTapButton(sender: UIButton)  {
```
어떤 버튼이 눌린건지 명확한 이름이 필요해 보입니다.

Wody95 commented on Expo1900/Expo1900/Controller/MainViewController.swift:
```diff
@@ -119,6 +135,8 @@ final class MainViewController: UIViewController {
+    // MARK: - Method
+    
     @objc func didTapButton(sender: UIButton)  {
         guard let secondView = self.storyboard?.instantiateViewController(withIdentifier: "ItemListView") as? ItemListViewController else { return }
```
navigationController에 push되는 viewController의 이름을 구체적으로 써주세요(물론 옆에 ItemListView가 써져있지만 변수명도 중요합니다)

Hugh-github committed:
docs: Step3 README 작성

seohyeon2 committed:
refactor: ItemTableViewCell 파일의 프로퍼티에 private 추가

seohyeon2 committed:
docs: README 파일에 UML 추가

Hugh-github committed:
refactor: Step3 리팩토링

- NavigationController의 강제 언랩핑 제거
- MainViewController의 오타 수정
- button의 titleLabel의 numberOfLines을 0으로 설정

seohyeon2 committed:
docs: README 파일의 UML 사진 수정

seohyeon2 committed:
refactor:  ExpoAssets 폴더안의 불필요한 파일 삭제

seohyeon2 commented:
> UIButton에도 폰트 크기에 따라 number Of line을 변경할 수 있으니 확인해보시면 좋을 것 같습니다. 
- `MainViewController`에 버튼에 `button.titleLabel?.numberOfLines = 0` 코드를 추가하여 해결했습니다! 알려주셔서 감사합니다😊

>현재 ExpoAssets 폴더 내부에 더 이상 사용하지 않는 imageSet이 있습니다. 현재 앱 내부의 이미지 소스는 Assets를 통해 관리하고 있으니 필요 없는 데이터는 지워주세요.
- 불필요한 파일 삭제 후 커밋했습니다!

seohyeon2 committed:
docs: README 파일안에 있는 이미지  크기 수정

seohyeon2 committed:
docs: README 파일 안의 표 2*2 형태로 변경
